common/app/concurrent/FutureSemaphore.scala

```diff
@@ -0,0 +1,24 @@
+package concurrent
+
+import java.util.concurrent.Semaphore
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Try}
+
+class FutureSemaphore(maxOperations: Int) {
+  private val semaphore = new Semaphore(maxOperations)
+
+  def execute[A](task: => Future[A])(implicit ec: ExecutionContext): Future[Try[A]] = {
+    if (semaphore.tryAcquire()) {
+      val resultF = task.map(Try(_)).recover { case e => Failure(e) }
+      resultF.foreach(_ => semaphore.release())
```

Comment:
a future is already including Failure/Success, you can use onComplete instead of recover then foreach to access both sides.

Reply:
I prefer to return a Future[Try[A]] to explicitly force the caller to do handle the failure case.

Reply:
I like that, use the type system to good advantage.

Reply:
I don't really see the point to introduce an intermediary `Try`. As John said `Future` already deal with failure case. And you cannot force the caller to deal with the error, as you proved it by simply calling `get` on the try.

Reply:
I don't want to enforce, just ensure the caller is aware. 

For my use it isn't necessary. I've only added it since FutureSemaphore is a reusable in `common`. Would you prefer to move FutureSemaphore out of `common`?